File: src/2/2_3/2_3_1/merge_sort.py
def merge(a, b):
    result = []
    while len(a) != 0 and len(b) != 0:
        element = a.pop(0) if a[0] < b[0] else b.pop(0)
        result.append(element)

    while len(a) != 0:
        result.append(a.pop(0))
    while len(b) != 0:
        result.append(b.pop(0))

    return result


def merge_sort(A):
    if len(A) <= 1:
        return A

    first_half = merge_sort(A[: len(A) // 2])
    second_half = merge_sort(A[len(A) // 2 :])
    return merge(first_half, second_half)

File: src/2/2.3/2.3-1/merge-sort.ts
function merge(a: Array<number>, b: Array<number>): Array<number> {
  const result = []

  while (a.length !== 0 && b.length !== 0) {
    const element = (a[0] < b[0] ? a.shift() : b.shift()) as number
    result.push(element)
  }

  return result.concat(a).concat(b)
}

export function mergeSort(elements: Array<number>): Array<number> {
  if (elements.length <= 1) return elements

  const midPoint = Math.ceil(elements.length / 2)
  const firstHalf = mergeSort(elements.slice(0, midPoint))
  const secondHalf = mergeSort(elements.slice(midPoint))
  return merge(firstHalf, secondHalf)
}
